### PR TITLE
Port C++ PREfast to C#

### PR DESCRIPTION
**Pull request by jeff-bacon**

### CODE REVIEW

File: src/Sarif.Converters/PREFastConverter.cs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                    }
+
+                    result.SetProperty(category, keyValuePair.Value);
+                }
+            }
+        }
+
+        private string RemoveLegacyNewLine(string content)
+        {
+            return content.Replace("PREFAST_NEWLINE\n", "");
```
Comment (jeff-bacon) [inline src/Sarif.Converters/PREFastConverter.cs]: PREfast files use this as their line break.  Remove it.

File: src/Sarif.Converters/PREFastObjectModel/Category.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+{
+    public class Category : Dictionary<string, string>, IXmlSerializable
+    {
+        public XmlSchema GetSchema()
+        {
+            return null;
+        }
+
+        public void ReadXml(XmlReader reader)
+        {
```
Comment (jeff-bacon) [inline src/Sarif.Converters/PREFastObjectModel/Category.cs]: Can't deserialize to a dictionary unless you tell it how

File: src/Sarif.FunctionalTests/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif [hunk 3]
```diff
@@ -9,22 +9,7 @@
       },
       "files": {
         "file:///d:/src/sarif-sdk/src/sarif.functionaltests/convertertestdata/prefast/src/advancedaliasing/aa_indexalias.cpp": {
-          "mimeType": "text/x-cpp",
-          "contents": "Ly8gVHJhY2sgImludGVyZXN0aW5nIiB2YXJpYWJsZXMgYW5kIGRldGVjdCB3aGVuIHRoZXkgYXJlIGFsaWFzZWQuDQoNCnZvaWQgQzYzODVfSW5kZXhBbGlhcyhpbnQgaW5wdXRJbmRleCkNCnsNCglpbnQgKiBjb250ZW50U3RvcmVzID0gbmV3IGludCBbMl07DQoNCgljb250ZW50U3RvcmVzWzBdID0gMDsNCg0KCWNvbnRlbnRTdG9yZXNbMV0gPSAxOw0KDQoJd2hpbGUgKGlucHV0SW5kZXggPiAxKQ0KCXsNCgkJaW50IGluZGV4ID0gaW5wdXRJbmRleDsNCg0KCQljb250ZW50U3RvcmVzWzBdID0gY29udGVudFN0b3Jlc1tpbmRleF07DQoJfQ0KDQoJZGVsZXRlIFtdIGNvbnRlbnRTdG9yZXM7DQp9DQo=",
```
Comment (jeff-bacon) [inline src/Sarif.FunctionalTests/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif]: A bunch of these probably need to be fixed.  What is the significance of this exactly?  All the tests pass without this being present.

**Comment by vinaykapadia** [comment]
>     // The functional tests for the PREfast converter must be handled differently from

Remove this comment, since PREfast is now C#.

---
Refers to: src/Sarif.FunctionalTests/SarifConverterTests.cs:22 in ce1f350. [](commit_id = ce1f350c83e396b71d96e4c77230654ed1ef9805, deletion_comment = False)

File: src/Sarif.FunctionalTests/SarifConverterTests.cs [hunk 4]
```diff
@@ -91,7 +91,7 @@ public void TSLint_EndToEnd()
             BatchRunConverter(ToolFormat.TSLint, "*.json", TestMode.CompareFileContents);
         }
 
-        [Fact(Skip = "BadImageFormatException introduced in net core port needs to be fixed")]
+        [Fact]
         public void PREfastConverter_EndToEnd()
         {
             BatchRunConverter(ToolFormat.PREfast, TestMode.CompareObjectModels);
```
Comment (vinaykapadia) [inline src/Sarif.FunctionalTests/SarifConverterTests.cs]: Does this still need to be compared this way? If it can be compared by CompareFileContents, maybe we should just do it that way all the time and remove the while TestMode stuff, to keep things clean?

**Comment by vinaykapadia** [comment]
This file hasn't changed in this PR so I can't comment on the file itself, so I did it here. In Sarif.Converters.UnitTests, BuiltInConverterFacoryTests.cs, there is a test called "BuiltInConverterFactory_HasConverterForEveryBuiltInToolFormatExceptPREfast". Since PREfast is now C# and is included in the built in converter factory, should probably change the test name, and also remove the ".Except(new[] { ToolFormat.PREfast })" part.

**Comment by vinaykapadia** [comment]
There are also a couple lines referencing PREfast in the appveyor.yml file. Don't know if these should be kept or removed, if we're no longer treating PREfast differently from the rest of the converters.

**Comment by jeff-bacon** [comment]
Fixed

---
In reply to: [339136843](https://github.com/Microsoft/sarif-sdk/pull/657#issuecomment-339136843) [](ancestors = 339136843)

File: src/Sarif.FunctionalTests/SarifConverterTests.cs (hunk 4 above)
Comment (jeff-bacon) [reply src/Sarif.FunctionalTests/SarifConverterTests.cs]: Done

---
In reply to: [146696755](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146696755) [](ancestors = 146696755)

**Comment by jeff-bacon** [comment]
>     // The functional tests for the PREfast converter must be handled differently from

Done

---
In reply to: [339135461](https://github.com/Microsoft/sarif-sdk/pull/657#issuecomment-339135461) [](ancestors = 339135461)

---
Refers to: src/Sarif.FunctionalTests/SarifConverterTests.cs:22 in ce1f350. [](commit_id = ce1f350c83e396b71d96e4c77230654ed1ef9805, deletion_comment = False)

File: src/Sarif.Converters.UnitTests/BuiltinConverterFactoryTests.cs [hunk 5]
```diff
@@ -12,10 +12,9 @@ namespace Microsoft.CodeAnalysis.Sarif
     public class BuiltInConverterFactoryTests
     {
         [Fact]
-        public void BuiltInConverterFactory_HasConverterForEveryBuiltInToolFormatExceptPREfast()
+        public void BuiltInConverterFactory_HasConverterForEveryBuiltInToolFormat()
```
Comment (lgolding) [inline src/Sarif.Converters.UnitTests/BuiltinConverterFactoryTests.cs]: LOL. Love this change (and that you thought to do it).

File: src/Sarif.FunctionalTests/SarifConverterTests.cs [hunk 6]
```diff
@@ -19,92 +19,67 @@
...
-    // formatting, than the other converters (which do use Newtonsoft.Json).
-    //
-    // Now, the functional tests work by deserializing a SARIF file into an object model,
-    // re-serializing it using Newtonsoft.Json, and comparing the two files. Obviously if
-    // the serialization and deserialization code doesn't match, the test will fail.
-    //
-    // So we introduce a "test mode" flag into the functional tests.For all tools except
-    // PREfast, we compare the serialized files.For PREfast, we compare the object models
-    // obtained by deserializing those files.
-    internal enum TestMode
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/SarifConverterTests.cs]: Good riddance!

File: src/Sarif.Converters/PREFastObjectModel/Category.cs (hunk 2 above)
Comment (lgolding) [reply src/Sarif.Converters/PREFastObjectModel/Category.cs]: If it's worth a comment in the PR review, it's worth a comment in the source code.

File: src/Sarif.Converters/PREFastObjectModel/Defect.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Xml.Serialization;
+
+namespace Microsoft.CodeAnalysis.Sarif.Converters.PREFastObjectModel
+{
+    public class Defect
+    {
+        public SFA SFA { get; set; }
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastObjectModel/Defect.cs]: Why is this here?

File: src/Sarif.FunctionalTests/ConverterTestData/PREfast/Standardc6306.cxx.xml.sarif [hunk 8]
```diff
@@ -29,6 +29,26 @@
       },
       "results": [
         {
+          "ruleId": "6273",
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/ConverterTestData/PREfast/Standardc6306.cxx.xml.sarif]: Why does the new converter produce a result that the old one didn't?

Comment (jeff-bacon) [reply src/Sarif.FunctionalTests/ConverterTestData/PREfast/Standardc6306.cxx.xml.sarif]: That's a good question.  When inspecting the original PREfast xml file, it does contain two defects.  It seems like this result is supposed to be there.  I don't see anything that would cause the C++ project to miss outputting this.

---
In reply to: [146913257](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146913257) [](ancestors = 146913257)

File: src/Sarif.Converters/PREFastObjectModel/Defect.cs (hunk 7 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastObjectModel/Defect.cs]: A PREfast defect has a child node named `SFA`.  I forgot to add `[XmlElement("SFA")]` to this but it seems to deserialize it anyway.

---
In reply to: [146912821](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146912821) [](ancestors = 146912821)

File: src/Sarif.Converters/PREFastObjectModel/Category.cs (hunk 2 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastObjectModel/Category.cs]: Fixed

---
In reply to: [146912420](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146912420) [](ancestors = 146912420)

File: src/Sarif.Converters/PREFastConverter.cs [hunk 9]
```diff
@@ -0,0 +1,193 @@
+﻿using System.IO;
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: Add copyright notice:
```
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

```

File: src/Sarif.Converters/PREFastConverter.cs [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+using Microsoft.CodeAnalysis.Sarif.Writers;
+using System;
+using System.Collections.Generic;
+using System.Xml;
+using System.Xml.Serialization;
+using Microsoft.CodeAnalysis.Sarif.Converters.PREFastObjectModel;
+using System.Linq;
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: Sort `using`s alphabetically (but with `System` `using`s first).

File: src/Sarif.Converters/PREFastConverter.cs [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+            var region = new Region
+            {
+                StartColumn = defect.SFA.Column + 1,
+                StartLine = defect.SFA.Line
+            };
+
+            Uri resultsFileUri = new Uri($"{defect.SFA.FilePath}{defect.SFA.FileName}", UriKind.Relative);
+            PhysicalLocation physicalLocation = new PhysicalLocation(uri: resultsFileUri, uriBaseId: null, region: region);
+            Location location = new Location()
+            {
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: You can use `var` for these three variables since the types are all explicit on the right-hand side.

File: src/Sarif.Converters/PREFastConverter.cs (hunk 10 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: Fixed

---
In reply to: [146948454](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146948454) [](ancestors = 146948454)

File: src/Sarif.Converters/PREFastConverter.cs (hunk 9 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: fixed

---
In reply to: [146948283](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146948283) [](ancestors = 146948283)

File: src/Sarif.Converters/PREFastConverter.cs (hunk 11 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: fixed

---
In reply to: [146949725](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146949725) [](ancestors = 146949725)

File: src/Sarif.Converters/PREFastConverter.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif.Converters/PREFastConverter.cs]: We generally use `string.Empty` instead of `""`.

File: src/Sarif.Converters/PREFastConverter.cs [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+using System.Xml;
+using System.Xml.Serialization;
+using Microsoft.CodeAnalysis.Sarif.Converters.PREFastObjectModel;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Sarif.Converters
+{
+    public class PREfastConverter : ToolFileConverterBase
+    {
+        private Dictionary<string, string> categories = new Dictionary<string, string>
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: Seeing how this is used below, should it be named something like "`knownCategories`"?

File: src/Sarif.Converters/PREFastConverter.cs [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+                Locations = locations
+            };
+
+            result.CodeFlows = new List<CodeFlow>()
+            {
+                codeFlow
+            };
+        }
+
+        private void ExtractCategories(Defect defect, Result result)
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: I think it should be called `SetCategories` in parallel with the other `Set` method names.

File: src/Sarif.Converters/PREFastConverter.cs [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+                    {
+                        category = categories[category];
+                    }
+
+                    result.SetProperty(category, keyValuePair.Value);
+                }
+            }
+        }
+
+        private string RemoveLegacyNewLine(string content)
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: I think "legacy" isn't quite the word here. How about `RemovePREfastNewline`?

File: src/Sarif.Converters/PREFastConverter.cs [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+            var codeFlow = new CodeFlow()
+            {
+                Locations = locations
+            };
+
+            result.CodeFlows = new List<CodeFlow>()
+            {
+                codeFlow
+            };
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: Or:
```
result.CodeFlows = new List<CodeFlow>
{
    new CodeFlow
    {
        Locations = locations
    }
};
```

File: src/Sarif.Converters/PREFastConverter.cs [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+            if (defect.Path?.SFAs.Any() == false)
+                return;
+
+            int step = 0;
+            var locations = new List<AnnotatedCodeLocation>();
+            bool pathUsesKeyEvents = defect.Path.SFAs.Any(x => !string.IsNullOrWhiteSpace(x?.KeyEvent?.Id));
+
+            foreach (var sfa in defect.Path.SFAs)
+            {
+                var reg = new Region()
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: `reg` => `region`

Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: fixed

---
In reply to: [146951857](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146951857) [](ancestors = 146951857)

File: src/Sarif.Converters/PREFastConverter.cs (hunk 15 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: fixed

---
In reply to: [146951526](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146951526) [](ancestors = 146951526)

File: src/Sarif.Converters/PREFastConverter.cs (hunk 14 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: fixed

---
In reply to: [146951280](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146951280) [](ancestors = 146951280)

File: src/Sarif.Converters/PREFastConverter.cs (hunk 13 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: fixed

---
In reply to: [146951094](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146951094) [](ancestors = 146951094)

File: src/Sarif.Converters/PREFastConverter.cs (hunk 12 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: fixed

---
In reply to: [146950815](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146950815) [](ancestors = 146950815)

File: src/Sarif.Converters/PREFastConverter.cs (hunk 1 above)
Comment (jeff-bacon) [reply src/Sarif.Converters/PREFastConverter.cs]: fixed

---
In reply to: [146681613](https://github.com/Microsoft/sarif-sdk/pull/657#discussion_r146681613) [](ancestors = 146681613)
